File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")
        pattern = list(pattern)
        hm = {}
        if len(pattern) != len(words):
            return False
        
        for i in range(len(pattern)):
            ch = pattern[i]
            word = words[i]
            if ch in hm:
                if hm[ch] != word:
                    return False
            elif word in set(hm.values()):
                return False
            else:
                hm[ch] = word
        
        return True

File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        HashMap<Character, String> hm = new HashMap<>();
        String[] words = s.split(" ");
        char ch;
        String word;
        
        if (pattern.length() != words.length) {
            return false;
        }
        
        for (int i = 0; i < pattern.length(); i++) {
            ch = pattern.charAt(i);
            word = words[i];
            if (hm.containsKey(ch)) {
                if (!hm.get(ch).equals(word)) {
                    return false;
                }
            } else if (hm.containsValue(word)) {
                return false;   
            } else {
                hm.put(ch, word);
            }
        }
        
        return true;
    }
}